feat(imgproc): introduce ExecutionStrategy and optimize threshold (~30x speedup)

nilaysrivastava opened the pull request:
Partially addresses #538

### Summary
This PR introduces the `ExecutionStrategy` API to give users runtime control over parallelization. As discussed in #538, not all operations benefit from Rayon's default parallelism.

To validate this, I implemented the API in `threshold.rs` as a pilot. Benchmarks show that **forcing serial execution** for this memory-bound operation results in a **~30x speedup** on standard resolutions (1080p), proving the need for this granularity.

### Changes
1.  **Added `ExecutionStrategy` Enum** in `parallel.rs`:
    * `Auto`: Defaults to Rayon global pool (existing behavior).
    * `Serial`: Forces execution on the current thread (no overhead).
    * `Fixed(n)`: Creates a local thread pool with `n` threads.
2.  **Updated `threshold_binary`**:
    * Updated signature to accept `strategy: ExecutionStrategy`.
    * Implemented logic to switch between `par_iter` and standard `iter`.
3.  **Updated Benchmarks**:
    * Updated `bench_threshold.rs` to support the new signature and compare strategies.

### Performance Analysis
Benchmarked on Apple M-series (1920x1080 image).

| Strategy | Time (Mean) | Notes |
| :--- | :--- | :--- |
| **Auto (Parallel)** | **987.48 µs** | High overhead for simple ops |
| **Serial** | **31.42 µs** | **~30x Faster** (Zero overhead) |

Benchmarks show that for memory-bound operations like thresholding, the overhead of parallelism outweighs the benefits. By switching to Serial execution, we bypass this overhead, achieving a ~30x speedup (reaching memory bandwidth limits) on standard 1080p images compared to the default parallel implementation.

This confirms that for lightweight, memory-bound operations like thresholding, the overhead of managing threads significantly outweighs the benefits. The `Serial` strategy allows users (and future default optimizations) to bypass this overhead.

### Checklist
- [x] Implementation of `ExecutionStrategy`
- [x] Applied to `threshold.rs`
- [x] Verified with `cargo test`
- [x] Verified with `cargo bench`

nilaysrivastava commented:
My apologies! I realised I staged the API definition but missed adding the implementation and benchmark files to the commit. I've just updated the PR with the full implementation.

nilaysrivastava commented on crates/kornia-imgproc/src/parallel.rs:
```diff
@@ -128,3 +140,41 @@ pub fn par_iter_rows_resample<const C: usize, A: ImageAllocator>(
+                let pool = rayon::ThreadPoolBuilder::new()
+                    .num_threads(n)
+                    .build()
+                    .expect("Failed to create thread pool");
```
I kept the panic! here to keep the API ergonomic. Since Fixed(n) is primarily for benchmarking/testing, forcing the user to unwrap a Result every time would be cumbersome. I did add a descriptive error message to the panic as suggested.

nilaysrivastava commented:
Refactor pushed! I've addressed the review comments:

- Added `ExecuteExt` trait abstraction to `parallel.rs`.
- Added validation for `Fixed(0)` and clearer panic messages.
- Added docs warning about thread pool overhead.
- Added benchmarks for `Fixed` strategy to demonstrate overhead.
- Updated PR description to 'Partially addresses #538'.

edgarriba commented on crates/kornia-imgproc/src/parallel.rs:
```diff
@@ -128,3 +151,99 @@ pub fn par_iter_rows_resample<const C: usize, A: ImageAllocator>(
+            }
+            ExecutionStrategy::Fixed(n) => {
+                if n == 0 {
+                    panic!("ExecutionStrategy::Fixed(n) requires n > 0");
```
no panics, return proper error via thiserror

edgarriba commented on crates/kornia-imgproc/src/parallel.rs:
```diff
@@ -128,3 +151,99 @@ pub fn par_iter_rows_resample<const C: usize, A: ImageAllocator>(
+                        panic!("Failed to create thread pool with {} threads: {}", n, e);
+                    });
+
+                pool.install(|| self.par_iter().zip(dst.par_iter_mut()).for_each(op));
```
return result

nilaysrivastava commented:
Updates:

- Renamed `AutoFull` to `ParallelElements` in `ExecutionStrategy` for clarity.
- Extended `ImageError` in `kornia-image` with a `Parallel` variant to properly propagate errors, removing the `expect()` panic.
- Fixed Benchmarks by moving the destination image allocation outside the benchmark loop to measure only the algorithm's performance.
- Added Tests for edge cases (invalid stride, zero thread count).
- Fixed the `rayon` version mismatch to use the workspace default.
- Removed all internal/agentic comments.
